### PR TITLE
infrastructure with blobfuse & containers is implemented

### PR DESCRIPTION
**Pull request by huseyingokay**

### CODE REVIEW

File: clone-project.sh [hunk 1]
```diff
@@ -1,21 +1,29 @@
-
 slug=$1
 sha=$2
+projdir=$(echo ${slug} | cut -d'-' -f1 | tr . /)
```
Comment (winglam) [inline clone-project.sh]: For this variable just have all scripts that call clone-project.sh pass in the ${modifiedslug_with_sha}=${modified_module}

This cut being done in this line wouldn't work if the slug contains '-'. E.g., github.com/some-user/some-repo

File: clone-project.sh [hunk 2]
```diff
@@ -1,21 +1,29 @@
...
+
+echo "in clone-project.sh"
+echo "slug: $slug"
+echo "sha: $sha"
+echo "projdir: $projdir"
```
Comment (winglam) [inline clone-project.sh]: Please remove these extra echo commands

File: clone-project.sh [hunk 3]
```diff
@@ -1,21 +1,29 @@
...
-if [[ ! -d "$slug" ]]; then
-    echo "================Cloning the project: $(date)"
-    git clone https://github.com/$slug $slug
-    cd $slug
+if [[ ! -f "$AZ_BATCH_TASK_WORKING_DIR/input/"$slug".zip" ]]; then
+    git clone https://github.com/$projdir $projdir
+    cd $AZ_BATCH_TASK_WORKING_DIR/$input_container/$projdir
     git checkout $sha
     echo "SHA is $(git rev-parse HEAD)"
 else
```
Comment (winglam) [inline clone-project.sh]: Originally this check would echo and cd into $slug if the project appears to be cloned already. 

We should add an if clause to check for such a case before either (1) cloning the project ourselves or (2) looking into the input container to see if the project was compiled before. Otherwise, this current code may have issues if a zip file from the input container was already unzipped

File: install-project.sh [hunk 4]
```diff
@@ -1,16 +1,23 @@
-
 slug=$1
-MVNOPTIONS=$2
+MVNOPTIONS="$2 -Dmaven.repo.local=$AZ_BATCH_TASK_WORKING_DIR/input/dependencies"
```
Comment (winglam) [inline install-project.sh]: Instead of changing the MVNOPTIONS here, you should directly change the MVNOPTIONS in run_surefire_azure.sh. The reason for this is because scripts like run_surefire_azure.sh makes many more calls to mvn later and they would all need -Dmaven.repo.local set as well.

On second thought though, can you check if there is a global argument that can be set once so that all subsequent calls to mvn would just use the correct -Dmaven.repo.local without needing to provide the argument? If you can globally set it once for mvn then you wouldn't need to change every script's MVNOPTIONS and you can just globally set it once in the run_azure.py script

File: run_surefire_rule_azure.sh [hunk 5]
```diff
@@ -34,7 +34,7 @@ short_sha=${sha:0:7}
 modifiedslug_with_sha="${modifiedslug}-${short_sha}"
 
 # echo "================Cloning the project"
-bash $dir/clone-project.sh "$slug" "$sha"
+bash $dir/clone-project.sh "${modifiedslug_with_sha}=${modified_module}" "$sha"
```
Comment (winglam) [inline run_surefire_rule_azure.sh]: The modified_module variable would need to be defined before this line. Currently, it is defined in Line 97

File: install-project.sh [hunk 6]
```diff
@@ -143,5 +150,10 @@ else
...
 fi
 
+cd ~/
+zip -r "${modifiedslug_with_sha}=${modified_module}".zip $projectname
+cp "${modifiedslug_with_sha}=${modified_module}".zip ~/input
+cd ~/$slug
+
 ret=${PIPESTATUS[0]}
-exit $ret
+exit $ret
```
Comment (winglam) [inline install-project.sh]: Please avoid these unnecessary changes by adding the line breaks back in this file and the other files

File: install-project.sh [hunk 7]
```diff
@@ -143,5 +150,10 @@ else
     mvn clean install -am -pl $module -DskipTests ${MVNOPTIONS} |& tee mvn-install.log
 fi
 
+cd ~/
```
Comment (winglam) [inline install-project.sh:155]: We should check if the zip already exists or not. If it does, we need not zip and copy the zip file again.

File: install-project.sh [hunk 8]
```diff
@@ -1,16 +1,23 @@
...
 fullTestName=$7
 RESULTSDIR=$8
+projectname=${slug%/*}
 
 modifiedslug=$(echo ${slug} | sed 's;/;.;' | tr '[:upper:]' '[:lower:]')
 short_sha=${sha:0:7}
 modifiedslug_with_sha="${modifiedslug}-${short_sha}"
+modified_module=$(echo ${module} | cut -d'.' -f2- | cut -c 2- | sed 's/\//+/g')
+
+if [[ ! -f "$AZ_BATCH_TASK_WORKING_DIR/input/"${modifiedslug_with_sha}=${modified_module}".zip" ]]; then
```
Comment (winglam) [inline install-project.sh]: Should we delete this if clause? This clause seems to prevent any project from actually compiling if it does not already exist in the input container. 

This script is also missing an if clause to prevent projects from being installed if the project is already in the input container.

File: install-project.sh [hunk 9]
```diff
@@ -1,16 +1,23 @@
...
 slug=$1
-MVNOPTIONS=$2
+MVNOPTIONS="$2 -Dmaven.repo.local=$AZ_BATCH_TASK_WORKING_DIR/input/dependencies"
 USER=$3
 module=$4
 sha=$5
 dir=$6
 fullTestName=$7
 RESULTSDIR=$8
+projectname=${slug%/*}
```
Comment (winglam) [inline install-project.sh]: We should be able to delete this variable and just directly use ${slug%/*} in Line 154

File: install-project.sh (hunk 8 above)
Comment (huseyingokay) [reply install-project.sh]: Ops, I put the "!" by mistake in the if statement, I am fixing

File: clone-project.sh (hunk 3 above)
Comment (huseyingokay) [reply clone-project.sh]: I don't know how this cd $slug was pushed but it is cd ~/

File: clone-project.sh (hunk 1 above)
Comment (huseyingokay) [reply clone-project.sh]: Yes you are right, I will remove that line but i need the original slug to be able to clone the project or go into the project folder. Thus, can I pass ${modifiedslug_with_sha}=${modified_module} and slug seperately to clone-project.sh?

Comment (winglam) [reply clone-project.sh]: You have slug and sha variable and two scenarios:
(1) A project does not exist in input containers - in this scenario just use slug and sha as the slug and sha
(2) A project does exist in input containers - in this scenario just use slug and sha as the slug and ${modifiedslug_with_sha}=${modified_module}

Would this work?

File: clone-project.sh (hunk 3 above)
Comment (winglam) [reply clone-project.sh]: The cd $slug existed before your changes. I don't believe you should change it.

Please also note that you are missing the else clause to check if a project has already been cloned or unzipped

File: clone-project.sh (hunk 1 above)
Comment (huseyingokay) [reply clone-project.sh]: I think it wouldn't because for example:
slug: codingchili/excelastic
sha: 6bb7884b1e70f9e588d75bdaf132a84f6e896a7a and I am checking the existance of the project in clone-project.sh. So if the project exists, then I can not create ${modifiedslug_with_sha}=${modified_module} with these two parameter. Am I missing something :(

Comment (winglam) [reply clone-project.sh]: We need not create ${modifiedslug_with_sha}=${modified_module} in clone-project.sh.

For (2), you can just pass in ${modifiedslug_with_sha}=${modified_module} instead of passing in sha. We can then have if statements to check whether sha is actually sha or if it is ${modifiedslug_with_sha}=${modified_module} instead. 

(In fact, sha should be renamed then to sha_or_modified_slug_sha_module)

Comment (huseyingokay) [reply clone-project.sh]: But in this version of sh files, I don't know if the project exists or not in the outside of the clone-project.sh. In run_surefire_rule_azure.sh, I am just giving two parameters; I am not checking if the file is exist or not in run_surefire_rule_azure.sh

Comment (winglam) [reply clone-project.sh]: Ah I see. Then I suggest two options: 
(1) Keep projdir but rename it to be modified_slug_sha_module and just pass in ${modifiedslug_with_sha}=${modified_module} whenever clone-project.sh is called, or
(2) Remove projdir, rename sha to be modified_slug_sha_module, and pas in ${modifiedslug_with_sha}=${modified_module}. You should then be able to retrieve the sha with the following. sha=$(echo $modified_slug_sha_module | rev | cut -d'=' -f2 | cut -d'-' -f1 | rev). Note that this command will give you the short sha but all in intents and purposes, the short sha will function the same as the long sha.

File: clone-project.sh (hunk 3 above)
Comment (huseyingokay) [reply clone-project.sh]: I couldn't get the idea of checking whether it is cloned or zipped. Is it for working on a same machine situation?

Comment (winglam) [reply clone-project.sh]: Yes. If this clone-project.sh is run twice, it should not clone or unzipped the project twice.

Comment (huseyingokay) [reply clone-project.sh]: line 6 d

Comment (huseyingokay) [reply clone-project.sh]: change line 6 -> cd ~/ to ~/AZ_BATCH...

Comment (huseyingokay) [reply clone-project.sh]: unzip -> unzip -q

Comment (huseyingokay) [reply clone-project.sh]: 23 -> remove git checkout, add enter at the end

**Review by winglam — COMMENTED** [review]
LGTM!
